cpp/src/arrow/compute/kernels/vector_take.cc

```diff
@@ -38,44 +55,715 @@ std::unique_ptr<KernelState> InitTake(KernelContext*, const KernelInitArgs& args
+    if (args.values_null_count > 0 || args.indices_null_count > 0) {
+      BitUtil::SetBitsTo(out_bitmap, out_offset, args.indices_length, false);
+    }
+
+    OptionalBitBlockCounter indices_bit_counter(indices_bitmap, indices_offset,
+                                                args.indices_length);
+    int64_t position = 0;
+    int64_t valid_count = 0;
+    while (true) {
+      BitBlockCount block = indices_bit_counter.NextBlock();
+      if (block.length == 0) {
+        // All indices processed.
+        break;
+      }
+      if (args.values_null_count == 0) {
+        // Values are never null, so things are easier
+        valid_count += block.popcount;
+        if (block.popcount == block.length) {
+          // Fastest path: neither values nor index nulls
+          BitUtil::SetBitsTo(out_bitmap, out_offset + position, block.length, true);
+          for (int64_t i = 0; i < block.length; ++i) {
+            out[position] = values[indices[position]];
+            ++position;
+          }
+        } else if (block.popcount > 0) {
+          // Slow path: some indices but not all are null
+          for (int64_t i = 0; i < block.length; ++i) {
+            if (BitUtil::GetBit(indices_bitmap, indices_offset + position)) {
+              // index is not null
+              BitUtil::SetBit(out_bitmap, out_offset + position);
+              out[position] = values[indices[position]];
+            }
+            ++position;
+          }
+        }
+      } else {
+        // Values have nulls, so we must do random access into the values bitmap
+        if (block.popcount == block.length) {
+          // Faster path: indices are not null but values may be
+          for (int64_t i = 0; i < block.length; ++i) {
+            if (BitUtil::GetBit(values_bitmap, values_offset + indices[position])) {
+              // value is not null
+              out[position] = values[indices[position]];
+              BitUtil::SetBit(out_bitmap, out_offset + position);
+              ++valid_count;
+            }
+            ++position;
+          }
+        } else if (block.popcount > 0) {
+          // Slow path: some but not all indices are null. Since we are doing
+          // random access in general we have to check the value nullness one by
+          // one.
+          for (int64_t i = 0; i < block.length; ++i) {
+            if (BitUtil::GetBit(indices_bitmap, indices_offset + position)) {
+              // index is not null
+              if (BitUtil::GetBit(values_bitmap, values_offset + indices[position])) {
+                // value is not null
+                out[position] = values[indices[position]];
+                BitUtil::SetBit(out_bitmap, out_offset + position);
+                ++valid_count;
+              }
+            }
+            ++position;
+          }
+        }
+      }
+    }
+    out_arr->null_count = out_arr->length - valid_count;
+  }
+};
+
+template <typename IndexCType>
+struct BooleanTakeImpl {
```

Comment:
TODO: I will add some random data unit tests for boolean type, which are only sparsely tested in the test suite